src/consensus/aft/raft.h

```diff
@@ -2316,103 +2157,14 @@ namespace aft
-            "Configurations: discard committed configuration at {}", conf->idx);
-          configurations.pop_front();
-          changed = true;
+        RAFT_DEBUG_FMT(
+          "Configurations: discard committed configuration at {}", conf->idx);
+        configurations.pop_front();
+        changed = true;
 
-          if (retired_node_cleanup && is_primary())
-          {
-            retired_node_cleanup->cleanup();
-          }
-        }
-        else
+        if (retired_node_cleanup && is_primary())
         {
-          if (
-            !is_retired() &&
-            conf->nodes.find(state->my_node_id) != conf->nodes.end() &&
-            next->nodes.find(state->my_node_id) == next->nodes.end())
-          {
-            if (!is_retiring())
-            {
-              become_retiring();
-            }
-            else
-            {
-              become_retired(idx, kv::RetirementPhase::Committed);
-            }
-          }
-
-          size_t num_trusted_nodes = num_trusted(*next);
-          size_t num_retired_nodes = num_retired(*conf, *next);
-          size_t num_required_retired_nodes =
-            num_required_retirements(*conf, *next);
-          if (
-            num_trusted_nodes == next->nodes.size() &&
-            num_retired_nodes == num_required_retired_nodes)
-          {
-            RAFT_TRACE_FMT(
-              "Configurations: all nodes trusted ({}) or retired ({}), "
-              "switching to configuration #{}",
-              num_trusted_nodes,
-              num_retired_nodes,
-              next->rid);
-
-            if (
-              is_learner() &&
-              next->nodes.find(state->my_node_id) != next->nodes.end())
-            {
-              RAFT_INFO_FMT(
-                "Becoming follower {}: {}",
-                state->my_node_id,
-                state->current_view);
-              leadership_state = kv::LeadershipState::Follower;
-              membership_state = kv::MembershipState::Active;
-            }
-
-            for (auto& [nid, _] : next->nodes)
-            {
-              learner_nodes.erase(nid);
-            }
-
-            for (auto& [nid, _] : conf->nodes)
-            {
-              if (next->nodes.find(nid) == next->nodes.end())
-              {
-                retired_nodes.erase(nid);
-              }
-            }
-
-            if (retired_node_cleanup && is_primary())
-            {
-              retired_node_cleanup->cleanup();
-            }
-
-            configurations.pop_front();
-          }
-          else
-          {
-            RAFT_TRACE_FMT(
-              "Configurations: not enough trusted or retired nodes for "
-              "configuration #{} ({}/{} trusted, {}/{} retired)",
-              next->rid,
-              num_trusted_nodes,
-              next->nodes.size(),
-              num_retired_nodes,
-              num_required_retired_nodes);
-            if (
-              node_client && !is_learner() &&
```

Comment:
I think we can remove the `node_client` member from raft too, as well as all mentions of the learner state?

Reply:
The `node_client` is used to by the RetiredNodeCleanup, and so cannot be removed.